src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledSpanBuilder`1.cs

```diff
@@ -0,0 +1,1345 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Microsoft.AspNetCore.Razor.Utilities;
+
+namespace Microsoft.AspNetCore.Razor.PooledObjects;
+
+/// <summary>
+///  Wraps a pooled array but doesn't allocate it until it's needed. Provides
+///  access to the backing array as a <see cref="Span{T}"/>.
+///  Note: Disposal ensures the pooled array is returned to the pool.
+/// </summary>
+[NonCopyable]
```

Comment:
Couldn't this be a ref struct rather than `[NonCopyable]`?

Reply:
Also, if it's `[NonCopyable]` and intended to be used in using statements, it needs the `AsRef()` extension method and the analyzer that supports `PooledArrayBuilder<T>`.